### ISSUE TITLE
Can Nanogui run without X11 (or Wayland)?

### ISSUE THREAD

**Issue by Fredrum** [mitsuba-renderer/nanogui#82]
Hi!
Yes so I'm looking for a gui solution that has GL surfaces and can run on Linux without X11.
So was wondering if nanogui can do that?

Cheers

**Comment by jeroen94704** [comment]
AFAIK, no, not at this time. At least not out-of-the-box. NanoGUI relies on GLFW for window creation, input events and getting an OpenGL context to render the GUI to. GLFW in turn requires X11 or Wayland to do this.

You could compile GLFW with OSMESA (Off-screen mesa) support, but then you loose hardware acceleration. There is also a pull-request for GLFW that makes it work with raw OpenGL (and hence does not need X11/Wayland), but this pull-request has been open for a couple of years now and is not yet merged.

**Comment by jeroen94704** [comment]
Do you keep the GPU acceleration when using SDL2? That's my main goal here. 

I've been experimenting with NanoGUI/GLFW on an arm64/Wayland/GLES2 platform, but I'm having a hard time getting it to work. I think Wayland/GLES support in GLFW is pretty spotty at this point. When I use the Wayland kiosk-shell even the basic GLES example in GLFW doesn't work (fails with some error message about the framebuffer being larger than the screen resolution).

So if NanoGUI with SDL2 works well I might give that a shot.

**Comment by RolandHughes** [comment]
> Do you keep the GPU acceleration when using SDL2? That's my main goal here.
> 
> I've been experimenting with NanoGUI/GLFW on an arm64/Wayland/GLES2 platform, but I'm having a hard time getting it to work. I think Wayland/GLES support in GLFW is pretty spotty at this point. When I use the Wayland kiosk-shell even the basic GLES example in GLFW doesn't work (fails with some error message about the framebuffer being larger than the screen resolution).
> 
> So if NanoGUI with SDL2 works well I might give that a shot.

Yes, Wayland/GLES under arm64 isn't full. It isn't full OpenGL either. You can see this question I had for Toradex about Nuklear a few days ago.
https://community.toradex.com/t/imx8-xopendisplay-null-weston-vivante/15480

I'm currently trying to deploy NanoGUI to a Toradex arm64 myself. Personally I'm hoping to use raw X11 without Wayland.

**Comment by jeroen94704** [comment]
The NanoGUI-SDL port mentioned above by @Fredrum works directly with an OpenGL ES context (created through SDL2). I got the NanoGUI demo running that way on an Apalis iMX8 with TorizonCore installed (after some modifications), and it does run full-screen. Still not sure about the hardware acceleration though.

**Comment by Fredrum** [comment]
@jeroen94704    If you look in the test code here https://pastebin.com/s9GsXg8f  there's commented out sections that will check what context and renderer your SDL2 session is using.  Would that be enough to tell you if your sessions are HW accellerated?  Maybe there's some OpenGL (ES) code you could add that would provide more?

Could I ask you if you are using 'nanogui-sdl'  or if you managed to get regualar nanogui to use the SDL2 context?

I'm having some a hard time with nanogui-sdl, figuring out how to do things without glitching and there seems to be a bunch of bugs that I'm spending a lot of time figuring out how to work around. I'm thinking it might be easier if I could get regular Nanogui to run with SDL2.
